rosetta/services/construction_service.go

```diff
@@ -0,0 +1,368 @@
+	"github.com/coinbase/rosetta-sdk-go/server"
+	"github.com/coinbase/rosetta-sdk-go/types"
+)
+
+type constructionAPIService struct {
+	elrondClient client.ElrondClientHandler
+	config       *configuration.Configuration
+	txsParser    *transactionsParser
+}
+
+// NewConstructionAPIService creates a new instance of an ConstructionAPIService.
+func NewConstructionAPIService(elrondClient client.ElrondClientHandler, cfg *configuration.Configuration) server.ConstructionAPIServicer {
+	return &constructionAPIService{
+		elrondClient: elrondClient,
+		config:       cfg,
+		txsParser:    newTransactionParser(cfg),
+	}
+}
+
+func checkOperationsAndMeta(ops []*types.Operation, meta map[string]interface{}) *types.Error {
+	terr := ErrConstructionCheck
+	if len(ops) == 0 {
+		terr.Message += "invalid number of operations"
+		return terr
+	}
+
+	for _, op := range ops {
+		if !checkOperationsType(op) {
+			terr.Message += "unsupported operation type"
+			return terr
+		}
+	}
+
+	if meta["gasLimit"] != nil {
+		if _, ok := meta["gasLimit"].(uint64); ok {
+			terr.Message += "invalid gas limit"
+			return terr
+		}
+	}
+	if meta["gasPrice"] != nil {
+		if _, ok := meta["gasPrice"].(uint64); ok {
+			terr.Message += "invalid gas price"
+			return terr
+		}
+	}
+
+	return nil
+}
+
+func checkOperationsType(op *types.Operation) bool {
+	for _, supOp := range SupportedOperationTypes {
+		if supOp == op.Type {
+			return true
+		}
+	}
+
+	return false
+}
+
+func getOptionsFromOperations(ops []*types.Operation) objectsMap {
+	options := make(objectsMap)
+	options["sender"] = ops[0].Account.Address
+	options["receiver"] = ops[1].Account.Address
+	options["type"] = ops[0].Type
+	options["value"] = ops[1].Amount.Value
+
+	return options
+}
+
+func (s *constructionAPIService) ConstructionPreprocess(
+	_ context.Context,
+	request *types.ConstructionPreprocessRequest,
+) (*types.ConstructionPreprocessResponse, *types.Error) {
+	if err := checkOperationsAndMeta(request.Operations, request.Metadata); err != nil {
+		return nil, err
+	}
+
+	options := getOptionsFromOperations(request.Operations)
+
+	if len(request.MaxFee) > 0 {
+		maxFee := request.MaxFee[0]
+		if maxFee.Currency.Symbol != s.config.Currency.Symbol ||
+			maxFee.Currency.Decimals != s.config.Currency.Decimals {
+			terr := ErrConstructionCheck
+			terr.Message += "invalid currency"
+			return nil, terr
+		}
+
+		options["maxFee"] = maxFee.Value
+	}
+
+	if request.SuggestedFeeMultiplier != nil {
+		options["feeMultiplier"] = *request.SuggestedFeeMultiplier
+	}
+
+	if request.Metadata["gasLimit"] != nil {
+		options["gasLimit"] = request.Metadata["gasLimit"]
+	}
+	if request.Metadata["gasPrice"] != nil {
+		options["gasPrice"] = request.Metadata["gasPrice"]
+	}
+	if request.Metadata["data"] != nil {
+		options["data"] = request.Metadata["data"]
+	}
+
+	return &types.ConstructionPreprocessResponse{
+		Options: options,
+	}, nil
+}
+
+func (s *constructionAPIService) ConstructionMetadata(
```

Comment:
add comment
change receiver name to `c` or `cas`

Reply:
done.